crates/extensions/c8y_firmware_manager/src/actor.rs

```diff
@@ -0,0 +1,756 @@
+            &C8yTopic::ChildSmartRestResponse(operation_entry.child_id.clone()).to_string(),
+        );
+        let installed_firmware_payload = format!(
+            "115,{},{},{}",
+            operation_entry.name, operation_entry.version, operation_entry.server_url
+        );
+        let installed_firmware_message =
+            MqttMessage::new(&c8y_child_topic, installed_firmware_payload);
+        self.message_box
+            .send(installed_firmware_message.into())
+            .await?;
+        Ok(())
+    }
+
+    fn remove_entry_from_active_operations(
+        &mut self,
+        operation_key: &OperationKey,
+    ) -> ActiveOperationState {
+        if let Some(operation_state) = self.active_child_ops.remove(operation_key) {
+            operation_state
+        } else {
+            ActiveOperationState::Pending
+        }
+    }
+
+    // The symlink path should be <tedge-data-dir>/file-transfer/<child-id>/firmware_update/<file_cache_key>
+    fn create_file_transfer_symlink(
+        &self,
+        child_id: &str,
+        file_cache_key: &str,
+        original_file_path: &Path,
+    ) -> Result<PathBuf, FirmwareManagementError> {
+        let file_transfer_dir_path = self.config.validate_and_get_file_transfer_dir_path()?;
+
+        let symlink_dir_path = file_transfer_dir_path
+            .join(child_id)
+            .join("firmware_update");
+        let symlink_path = symlink_dir_path.join(file_cache_key);
+
+        if !symlink_path.is_symlink() {
+            fs::create_dir_all(symlink_dir_path)?;
+            unix_fs::symlink(original_file_path, &symlink_path)?;
+        }
+        Ok(symlink_path)
+    }
+
+    // Candidate to be removed since another actor should be in charge of this.
+    async fn get_pending_operations_from_cloud(&mut self) -> Result<(), FirmwareManagementError> {
+        let message = MqttMessage::new(&C8yTopic::SmartRestResponse.to_topic()?, "500");
+        self.message_box.send(message.into()).await?;
+        Ok(())
+    }
+}
+
+#[async_trait]
+impl Actor for FirmwareManagerActor {
+    fn name(&self) -> &str {
+        "FirmwareManager"
+    }
+
+    async fn run(&mut self) -> Result<(), RuntimeError> {
+        self.resend_operations_to_child_device().await?;
+        // TODO: We need a dedicated actor to publish 500 later.
+        self.get_pending_operations_from_cloud().await?;
+
+        info!("Ready to serve the firmware request.");
+        while let Some(event) = self.message_box.recv().await {
+            match event {
+                FirmwareInput::MqttMessage(message) => {
+                    self.process_mqtt_message(message).await?;
+                }
+                FirmwareInput::OperationTimeout(timeout) => {
+                    self.process_operation_timeout(timeout).await?;
+                }
+                FirmwareInput::IdDownloadResult((id, result)) => {
+                    self.process_after_download(&id, result).await?
+                }
+            }
```

Comment:
I would put the `impl Actor for FirmwareManagerActor` on top of this file before the `impl FirmwareManagerActor`, as this `run()` method is the entry point to understand the code.